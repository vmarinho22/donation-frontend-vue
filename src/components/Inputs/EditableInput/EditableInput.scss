.editable-input {
  width: 100%;
  display: flex;
  align-items: center;
  gap: 8px;
}