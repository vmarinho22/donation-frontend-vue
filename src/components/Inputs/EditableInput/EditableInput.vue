<template>
  <div class="editable-input">
    <p v-if="componentState === ComponentStateEnum.TEXT">{{ showValue }}</p>
    <div v-else>
      <slot></slot>
    </div>
    <v-btn
      variant="text"
      :icon="componentState === ComponentStateEnum.TEXT ? 'mdi-pencil-outline' : 'mdi-close'"
      size="small"
      @click="toggleState"
    />
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue'

enum ComponentStateEnum {
  TEXT = 'text',
  EDITABLE = 'editable'
}

type ComponentState = ComponentStateEnum.TEXT | ComponentStateEnum.EDITABLE

const { showValue, defaultState = ComponentStateEnum.TEXT } = defineProps<{
  showValue: string | number
  defaultState?: ComponentState
}>()

const componentState = ref<string>(defaultState)

function toggleState() {
  componentState.value =
    componentState.value === ComponentStateEnum.TEXT
      ? ComponentStateEnum.EDITABLE
      : ComponentStateEnum.TEXT
}
</script>

<style scoped lang="scss" src="./EditableInput.scss" />
