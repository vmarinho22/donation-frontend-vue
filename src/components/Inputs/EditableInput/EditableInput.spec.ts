import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import EditableInput from './EditableInput.vue'
import vuetify from '@/tests/plugins/vuetify'

describe('EditableInput', () => {
  it('renders the component', () => {
    const wrapper = mount(EditableInput, {
      global: {
        plugins: [vuetify]
      },
      props: {
        showValue: 'test',
      },
      slots: {
        default: '<input type="text" />'
      }
    })
    expect(wrapper.html()).toMatchSnapshot();
  })

  it('should click on editable button', async () => {
    const wrapper = mount(EditableInput, {
      global: {
        plugins: [vuetify]
      },
      props: {
        showValue: 'test',
      },
      slots: {
        default: '<input type="text" />'
      }
    })

    const button = wrapper.find('.mdi-pencil-outline')

    await button.trigger('click')

    const input = wrapper.find('input')

    expect(wrapper.html()).toMatchSnapshot()
    expect(input.attributes('type')).toBe('text')
  })
})