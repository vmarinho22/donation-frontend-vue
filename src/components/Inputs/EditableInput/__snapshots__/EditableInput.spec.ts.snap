// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableInput > renders the component 1`] = `
"<div data-v-97845744=\\"\\" class=\\"editable-input\\">
  <p data-v-97845744=\\"\\">test</p><button data-v-97845744=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--icon v-theme--light v-btn--density-default v-btn--size-small v-btn--variant-text\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
    <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-pencil-outline mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
    <!---->
    <!---->
  </button>
</div>"
`;

exports[`EditableInput > should click on editable button 1`] = `
"<div data-v-97845744=\\"\\" class=\\"editable-input\\">
  <div data-v-97845744=\\"\\"><input type=\\"text\\"></div><button data-v-97845744=\\"\\" type=\\"button\\" class=\\"v-btn v-btn--icon v-theme--light v-btn--density-default v-btn--size-small v-btn--variant-text\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
    <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\"><i class=\\"mdi-close mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i></span>
    <!---->
    <!---->
  </button>
</div>"
`;
